models/spin_chain.py

```diff
@@ -0,0 +1,266 @@
+                # TODO: find simple way to turn PBC on/off, see 'nn' below
+                # Nearest Neighbor (NN) interactions
+                if alpha == 'nn':
+                    if callable(strength):
+                        graph = [[lambda t, s=strength, i=i: s(t, i,(i + 1) % L),
+                                  i, (i + 1) % L]for i in range(L)]
+                    else:
+                        graph = [[lambda t, s=strength: s, i, (i + 1) % L]
+                                 for i in range(L)]
+
+                # Next-Nearest Neighbor (NNN) interactions
+                elif alpha == 'nnn':
+                    if callable(strength):
+                        graph = [[lambda t, s=strength, i=i: s(t, i, i + 2), i,
+                                  i + 2]
+                                 for i in range(L - 1)]
+                    else:
+                        graph = [[lambda t, s=strength: s, i, i + 2]
+                                 for i in range(L - 1)]
+
+                else:
+                    raise ValueError(f"Invalid range cutoff string: {alpha}")
+
+                # Add to the dictionary for this operator
+                new_interaction_dict[operator].extend(graph)
+
+            elif alpha == np.inf:
+                if len(operator) != 1:
+                    raise ValueError(f"One-site operation requires one-site "
+                                     f"operator: {operator}")
+
+                # On-site interaction
+                if callable(strength):
+                    # Time-dependent or site-specific on-site term
+                    graph = [[lambda t, s=strength, i=i: s(t, i), i]
+                             for i in range(L)]
+                else:
+                    # Constant on-site term
+                    graph = [[lambda t, s=strength: s, i]
+                             for i in range(L)]
+
+                # Add to the dictionary for this operator
+                new_interaction_dict[operator].extend(graph)
+
+            else:
+                if len(operator) != 2:
+                    raise ValueError(f"Two-site operation requires two-site "
+                                     f"operator: {operator}")
+
+                # General long-range interaction
+                if callable(strength):
+                    # Time and site-dependent strength, inverse range alpha
+                    graph = [[lambda t, s=strength, i=i, j=j:
+                              s(t, i, j)/(np.abs(i-j)**alpha), i, j]
+                             for i in range(L) for j in range(L)]
+                else:
+                    # Constant interaction strength, inverse range alpha
+                    graph = [[lambda t, s=strength:
+                              s/(np.abs(i-j)**alpha), i, j]
+                             for i in range(L) for j in range(L)]
+
+                # Add to the dictionary for this operator
+                new_interaction_dict[operator].extend(graph)
```

Comment:
this logic looks repetitive -- probably refactorable? for example, it looks like you can handle the two ways to specify `strength` as
```python
strength_func = strength if callable(strength) else (lambda t, i, j: strength)
```

Reply:
Tried it, doesn't work